File: backend/src/docker-images/python/executor.py
import sys
import os
import io
import contextlib
import signal

# Configuration
EXECUTION_TIMEOUT = 3 # seconds
MAX_OUTPUT_LENGTH = 10000 # 10KB

def handler(signum, frame):
    raise TimeoutError("Time Limit Exceeded")

# Set timeout handler
signal.signal(signal.SIGALRM, handler)

def read_file(path):
    try:
        with open(path, 'r') as f:
            return f.read()
    except Exception as e:
        sys.stderr.write(f"Error reading solution file: {e}\n")
        sys.exit(1)

def execute_code(code, input_data):
    # Capture stdout/stderr
    stdout_capture = io.StringIO()
    stderr_capture = io.StringIO()
    
    # Prepare execution environment
    # We restrict builtins slightly, but mostly rely on Docker for isolation
    safe_globals = {
        '__builtins__': __builtins__,
        'input': lambda: input_data_iter.__next__() if input_data_iter else "",
        'print': lambda *args, **kwargs: print(*args, file=stdout_capture, **kwargs)
    }
    
    # Handle input iteration
    input_lines = input_data.split('\n')
    input_data_iter = iter(input_lines)
    
    try:
        # Start timeout
        signal.alarm(EXECUTION_TIMEOUT)
        
        with contextlib.redirect_stdout(stdout_capture), contextlib.redirect_stderr(stderr_capture):
            exec(code, safe_globals)
            
        # Stop timeout
        signal.alarm(0)
        
        # Process output
        output = stdout_capture.getvalue()
        error = stderr_capture.getvalue()
        
        if len(output) > MAX_OUTPUT_LENGTH:
            output = output[:MAX_OUTPUT_LENGTH] + "\n...[Output Truncated]"
            
        sys.stdout.write(output)
        if error:
            sys.stderr.write(error)
            
        sys.exit(0)
        
    except TimeoutError:
        sys.stderr.write("Time Limit Exceeded\n")
        sys.exit(124)
    except Exception as e:
        sys.stderr.write(f"Runtime Error: {e}\n")
        sys.exit(1)
        
if __name__ == "__main__":
    # Read user code
    user_code = read_file('/app/temp/solution.py')
    
    # Read test input from stdin
    test_input = sys.stdin.read()
    
    execute_code(user_code, test_input)

File: backend/src/docker-images/javascript/executor.js
#!/usr/bin/env node

const vm = require('vm');
const fs = require('fs');

/**
 * Secure JavaScript Code Executor
 * Runs user code in an isolated VM sandbox with strict limits
 */

// Configuration
const EXECUTION_TIMEOUT = 3000; // 3 seconds
const MAX_OUTPUT_LENGTH = 10000; // 10KB

// Read user code from file
let userCode = '';
try {
    userCode = fs.readFileSync('/app/temp/solution.js', 'utf8');
} catch (error) {
    process.stderr.write('Error reading solution file\n');
    process.exit(1);
}

// Read test input from stdin
let testInput = '';
process.stdin.setEncoding('utf8');

process.stdin.on('data', (chunk) => {
    testInput += chunk;
});

process.stdin.on('end', () => {
    executeCode(userCode, testInput.trim());
});

function executeCode(code, input) {
    // Output buffer
    let output = '';
    let errorOutput = '';

    // Create secure sandbox
    const sandbox = {
        // Limited console
        console: {
            log: (...args) => {
                const line = args.map(arg => String(arg)).join(' ') + '\n';
                if (output.length + line.length < MAX_OUTPUT_LENGTH) {
                    output += line;
                }
            },
            error: (...args) => {
                const line = args.map(arg => String(arg)).join(' ') + '\n';
                errorOutput += line;
            },
        },

        // Provide input
        input: input,

        // Safe JSON
        JSON: JSON,

        // Blocked globals
        require: undefined,
        process: undefined,
        global: undefined,
        Buffer: undefined,
    };

    try {
        // Execute with timeout
        vm.runInNewContext(code, sandbox, {
            timeout: EXECUTION_TIMEOUT,
            breakOnSigint: true,
            displayErrors: true,
        });

        // Write output to stdout
        process.stdout.write(output);

        if (errorOutput) {
            process.stderr.write(errorOutput);
        }

        process.exit(0);

    } catch (error) {
        // Handle execution errors
        if (error.code === 'ERR_SCRIPT_EXECUTION_TIMEOUT') {
            process.stderr.write('Time Limit Exceeded\n');
            process.exit(124); // Timeout exit code
        } else if (error.code === 'ERR_SCRIPT_EXECUTION_INTERRUPTED') {
            process.stderr.write('Execution Interrupted\n');
            process.exit(130); // Interrupted exit code
        } else {
            process.stderr.write(`Runtime Error: ${error.message}\n`);
            process.exit(1);
        }
    }
}

// Handle uncaught errors
process.on('uncaughtException', (error) => {
    process.stderr.write(`Uncaught Exception: ${error.message}\n`);
    process.exit(1);
});

process.on('unhandledRejection', (reason) => {
    process.stderr.write(`Unhandled Rejection: ${reason}\n`);
    process.exit(1);
});
